pkg/receive/hashring.go

```diff
@@ -108,7 +119,53 @@ func NewHashring(matcher Matcher, groups []*targetgroup.Group) Hashring {
 		matcher:   matcher,
 	}
 	for _, g := range groups {
-		m.hashrings[g.Source] = &simpleHashring{*g}
+		l, ok := g.Labels[tenantLabel]
+		if !ok {
+			l = ""
+		}
+		t := string(l)
+		if _, ok := m.hashrings[t]; !ok {
+			m.hashrings[t] = &simpleHashring{}
+		}
+		h := m.hashrings[t].(*simpleHashring)
+		h.Targets = append(h.Targets, g.Targets...)
 	}
 	return m
 }
+
+// HashringsAreEquivalent indicates whether the hashrings created from
+// two sets of groups would be equivalent.
+func HashringsAreEquivalent(a, b []*targetgroup.Group) bool {
```

Comment:
Hm what do you mean? If I understand you, we are doing exactly that here: we save the target group that produces a hashring, then compare it deeply to a new target group to see if they would produce equivalent rings. Note that doing a shallow comparison won’t work because every `interval` seconds the file is re-read, a new target group is generated which would never be the same as the original.

Reply:
I'm saying, maybe comparing targetGroups themselves would be simpler?

Reply:
Ah sorry, despite what the name says, this function accepts two `[]*targetgroup.Group` as its parameters